spec.html

```diff
@@ -120,11 +120,12 @@ <h1>AggregateError ( _errors_, _message_ )</h1>
         <emu-alg>
           1. If NewTarget is *undefined*, let _newTarget_ be the active function object, else let _newTarget_ be NewTarget.
           1. Let _O_ be ? OrdinaryCreateFromConstructor(_newTarget_, `"%AggregateError.prototype%"`, « [[ErrorData]], [[AggregateErrors]] »).
-          1. Let _errorsList_ be ? IterableToList(_errors_).
-          1. Set _O_.[[AggregateErrors]] to _errorsList_.
```

Comment:
This was very intentional so it matches the order of the arguments. There is no convention for argument ordering in JS to sort “superclass” things above “subclass” things, but there is a very strong one for “argument order”.

Reply:
In that case, we can reduce this patch to "2. Unify handling of the message parameter with Error ctor". @marjakh WDYT?

Reply:
@ljharb Consistency with argument order here will result in a deeper inconsistency in the mental model and, as @mathiasbynens has pointed out, inconsistency with the parameter handling of Error ctor.

Like other NativeErrors, AggregateError.prototype's [[Prototype]] is Error.prototype, so the subclass relation is already there and it would be natural for users to think of AggregateError as a subclass of Error. For user-written subclasses, constructors would have a `super(message)` call at the top, and so the superclass's arguments *would* get processed before the subclass's arguments. It'd be very odd if AggregateError wasn't observationally equivalent to the simple implementation.

(This was @bakkot's original design in https://github.com/tc39/proposal-promise-any/issues/14, the ordering got buggy in translation to ecmarkup.)

Per #14 above there is also a good reason the argument order is `errors, message` instead of `message, errors`: https://github.com/tc39/proposal-promise-any/issues/14#issuecomment-468122531. AggregateError is a thing that is useful when you have... an aggregate of errors, not when you just have an error message, and so `message` should come second.

tl;dr:
1. There is a convention that superclass fields be *processed* before subclass fields.
1. Argument order should err towards being ergonomic, not reflect class hierarchies or other implementation details.

All in all, this PR is the right fix.

Reply:
> For user-written subclasses, constructors would have a super(message) call at the top, and so the superclass's arguments would get processed before the subclass's arguments.

👍 

> Per #14 above there is also a good reason the argument order is errors, message instead of message, errors: #14 (comment). AggregateError is a thing that is useful when you have... an aggregate of errors, not when you just have an error message, and so message should come second.

This still feels wrong from a user perspective. Every error we provide receives its `message` first. If we want to support receiving only the `errors` array, we could be [juggling](https://github.com/tc39/proposal-promise-any/issues/44#issuecomment-614856554).

Reply:
@jridgewell The argument order is an orthogonal discussion (though it's also decision that has already been discussed at length), and I'd like to scope this PR to just the order of when arguments are processed.

For this PR I'm assuming the current API decision is upheld, and am arguing that even given this order, subclass consistency is the salient consistency, not argument order. Is that convincing?

Reply:
Yes, I think this PR is fine as is. I tried to separate my complaint on arg ordering from the overall processing change in my comment.

Reply:
I think we have two conflicting positions here.

Altho i prefer `errors, message`, i think that the argument order strongly determines the evaluation order (including exceptions). If we want to change the order, let’s change the order.

Reply:
> In that case, we can reduce this patch to "2. Unify handling of the message parameter with Error ctor". @marjakh WDYT?

My thinking is similar to syg@ (the comment below yours). Especially I find this pseudo code from #14 compelling:

```
class AggregateError extends Error {
  errors;
  constructor(errors, message) {
    super(message);
    this.errors = Array.from(errors);
  }
}
```
Also from implementor's point of view: we already have a function that's essentially the "super(message);", and it's weird that we cannot use it. It would be intuitive to implement this feature by reusing the Error construction code.

Reply:
The implementation should be:
```js
class AggregateError extends Error {
  #AggregateErrors;
  constructor(errors, message) {
    const errorsList = IterableToList(errors); // throws if needed
    super(message);
    this.#AggregateErrors = errorsList;
  }
}
```

I'm not sure why this makes "super(message)" any harder to reuse.

If there's an implementability reason why `errors` must be evaluated second, then I'd see that as an argument to invert the argument order, not as an argument to deviate from the almost universal convention (including in WebIDL, which i believe a number of people are hoping to convert 262 to use one day) of validating all arguments in the order they're passed.

Reply:
That version of course makes "super(message)" possible to reuse, but it's surprising that the super call is not the first line.

To summarize:

Option 1 (current):
```
constructor(errors, message) {
    super(undefined);
    const errorsList = IterableToList(errors);
    this.#AggregateErrors = errorsList;
    this.message = ToString(message);
  }
```
Downside: can't reuse super(message)

Option 2 (this PR):

```
constructor(errors, message) {
    super(message);
    const errorsList = IterableToList(errors);
    this.#AggregateErrors = errorsList;
  }
```

Downside: arguments not processed in order

Option 3:
```
constructor(errors, message) {
    const errorsList = IterableToList(errors);
    super(message);
    this.#AggregateErrors = errorsList;
  }
```

Downside: super() not the first line

Option 4:

```
constructor(message, errors) {
    super(message);
    const errorsList = IterableToList(errors);
    this.#AggregateErrors = errorsList;
  }
```

Downside: argument order not ergonomic

If I understand correctly, @ljharb would prefer:
1 > 3 > 4 > 2

syg@ and I prefer:
2 > {1, 3} > 4 (I'm undecided which one is more surprising, 3 or 1, and idk what syg@ thinks.)

Reply:
I have no expectation that super is the first line, and I’m not sure why that would be expected. The super call merely needs to happen prior to the first reference to the receiver.

Reply:
Also, to be clear, i don’t think 1 makes sense regardless - I’d prefer 3, or 4.

Reply:
Thanks for the clarification of your preferences. Right, maybe it's not too surprising that super() is not the first line (it doesn't have to be, but it's conventional that it is).

I would be happy with option 3 as well, let's wait what @syg says. Maybe that's the "lesser evil", as all of the options have downsides.

Reply:
@marjakh's right about my preferences. I do like 2 > 3 > 1. I'm going to leave 4 out of it because that hinges on an orthogonal discussion on what the most useful argument order is.

@ljharb I stand by my and other folks' preference for 2 here.

I simply don't get the sense that other folks share your very strong adherence to processing arguments in the order that they're presented in, and. And while you might not have the expectation that `super` is the first line, but that is *by far* the expectation across OO languages with inheritance. C++ and Java, for instance, enforce that by design. JS is more flexible here, but that doesn't invalidate the expectation that `super` is the first line. I'd like to see more than personal conviction that argument processing must be tied to order to invalidate this expectation.

Argument order in general do not suggest anything about how the arguments are consumed. I am baffled by this stance.